lua_modules/ftp/ftpserver.lua

```diff
@@ -0,0 +1,172 @@
+    end
+    if a1 == "MODE" then
+      if a2 ~= "S" then
+        return c:send("504 Only S(tream) is suported\r\n")
+      end
+      return c:send("200 S OK\r\n")
+    end
+    if a1 == "PASV" then
+      local _,ip = c:getaddr()
+      local _,_,i1,i2,i3,i4 = string.find(ip,"(%d+).(%d+).(%d+).(%d+)")
+      return c:send("227 Entering Passive Mode ("..i1..","..i2..","..i3..","..i4..",0,20).\r\n")
+    end
+    if a1 == "LIST" or a1 == "NLST" then
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        local l = file.list();
+        for k,v in pairs(l) do
+          sd:send( a1 == "NLST" and k.."\r\n" or "+r,s"..v..",\t"..k.."\r\n")
+        end
+        sd:close()
+        data_fnc = nil
+        c:send("226 Transfer complete.\r\n")
+      end
+      if data_sock then
+        node.task.post(function() data_fnc(data_sock);data_sock=nil end)
+      end
+      return
+    end
+    if a1 == "RETR" then
+      local f = file.open(a2:gsub("%/",""),"r")
+      if f == nil then
+        return c:send("550 File "..a2.." not found\r\n")
+      end
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        sd:on("sent", function(cd)
+          b=f:read(1024)
+          if b then
+            cd:send(b)
+            b=nil
+          else
+            cd:close()
+            f:close()
+            data_fnc = nil
+            c:send("226 Transfer complete.\r\n")
+          end
+        end)
+        local b=f:read(1024)
+        sd:send(b)
+        b=nil
+      end
+      if data_sock then
+        node.task.post(function() data_fnc(data_sock);data_sock=nil end)
+      end
+      return
+    end
+    if a1 == "STOR" then
+      local f = file.open(a2:gsub("%/",""),"w")
+      if f == nil then
+        return c:send("451 Can't open/create "..a2.."\r\n")
+      end
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        sd:on("receive", function(cd, dd)
+          f:write(dd)
+        end)
+        sd:on("disconnection", function(c)
+          f:close()
+          data_fnc = nil
+        end)
+        c:send("226 Transfer complete.\r\n")
+      end
+      if data_sock then
+        node.task.post(function() data_fnc(data_sock);data_sock=nil end)
+      end
+      return
+    end
+    if a1 == "RNFR" then
+      buf = a2
+      return c:send("350 RNFR accepted\r\n")
+    end
+    if a1 == "RNTO" and buf ~= "" then
+      file.rename(buf, a2)
+      buf = ""
+      return c:send("250 File renamed\r\n")
+    end
+    if a1 == "DELE" then
+      if a2 == nil or a2 == "" then
+        return c:send("501 No file name\r\n")
+      end
+      file.remove(a2:gsub("%/",""))
+      return c:send("250 Deleted "..a2.."\r\n")
+    end
+    if a1 == "NOOP" then
+      return c:send("200 OK\r\n")
+    end
+    if a1 == "QUIT" then
+      return c:send("221 Goodbye\r\n", function (s) s:close() end)
+    end
+    c:send("500 Unknown error\r\n")
```

Comment:
Add support for SIZE command: this is generally used by web-browsers.
~~~lua
    if a1 == "SIZE" then
      local st,size
      st = file.stat(string.sub(a2, 2))
      if st then
        return c:send("213 "..st.size.."\r\n")
      else
        return c:send("550 Could not get file size.\r\n")
      end
    end
~~~